graphistry/embed_utils.py

```diff
@@ -0,0 +1,593 @@
+                    (result, o), axis=0
+            ) if result is not None else o
+
+        result_df = []
+        for i in result:
+            s, r, d = i
+            result_df += [
+                    [
+                        self._id2node[s],
+                        self._id2relation[r],
+                        self._id2node[d]
+                    ]
+            ]
+        result_df = pd.DataFrame(
+                result_df,
+                columns=[src, rel, "predicted_destination"]
+        )
+
+        return result_df
+
+    def predict_links(
+        self,
+        threshold=0.99,
+        return_embeddings=True,
+        retain_old_edges=False
+    ):
+        """predict links over entire graph given a threshold
+
+        Args:
+            threshold (float, optional): Probability threshold.
+                Defaults to 0.99.
+            return_embeddings (bool, optional): will return DataFrame of
+                predictions and node embeddings. Defaults to True.
+            retain_old_edges (bool, optional): will include old edges in
+                predicted graph. Defaults to False.
+
+        Returns:
+            graph: (graphistry graph) or
+                    (graphistry graph, DataFrame of predictions and
+                    node embeddings) when return_embeddings=True
+        """
+
+        predicted_links, node_embeddings = self._predict(
+                    torch.tensor(self.triplets),
+                    threshold,
+                    infer="all"
+        )
+
+        existing_links = self._edges[
+                [self._source, self._relation, self._destination]
+        ]
+
+        if retain_old_edges:
+            all_links = pd.concat(
+                [existing_links, predicted_links],
+                ignore_index=True
+            ).drop_duplicates()
+        else:
+            all_links = predicted_links
+
+        g_new = self.nodes(
+                    self._nodes,
+                    self._node
+                )
+        g_new = g_new.edges(
+                    all_links,
+                    self._source,
+                    self._destination
+        )
+
+        if return_embeddings:
+            return g_new, predicted_links, node_embeddings
+        return g_new
+
+    def _score(self, triplets):
+        emb = torch.tensor(self._embeddings)
+        if type(triplets) != torch.Tensor:
+            triplets = torch.tensor(triplets)
+        score = self._embed_model.score(emb, triplets)
+        prob = torch.sigmoid(score)
+        return prob.detach()
+
+    def _eval(self, threshold):
+        if self.test_idx != []:
+            from time import time
+            t = time()
+            print("Evaluating...")
+            triplets = torch.tensor(self.triplets)[self.test_idx]
+            score = self._score(triplets)
+            score = 100 * len(score[score >= threshold]) / len(score)
+            print(f"--took {(time()-t)/60:.2f} minutes to evaluate")
```

Comment:
logger

Reply:
for this time being I am thinking about removing the eval time, and later it will be added in the pbar to make everything look a bit nicer